src/main/java/org/basseur/adventofcode/advent2018/days/day07/Day07.java

```diff
@@ -0,0 +1,220 @@
+    }
+
+    @Override
+    public String firstPart() {
+        return "Part 1 - Order in which the steps in the instructions should be completed: " + determineOrder();
+    }
+
+    @Override
+    public String secondPart() {
+        return "Part 2 - Time required to complete all of the steps: " + determineTime() + " seconds";
+    }
+
+    @Override
+    public HashMap<String, ProblemStatusEnum> getProblemStatus() {
+        return problemStatus;
+    }
+
+    /**
+     * Primary Method for Day 7, Part 1.
+     * <p>
+     * Determines the order, in which the instructions should be completed.
+     * For this purpose {@link #stepsMap} is copied into a list. The new HashMap is
+     * scanned for all {@link Step}s that have no previous steps in them. These
+     * are added to an ArrayList which then gets sorted alphabetically. The first
+     * item is the current step, which gets added to the output. Subsequently, this
+     * step gets removed from all {@link Step#previousSteps} and from the HashMap,
+     * as well as from the available Steps. This process gets repeated until the
+     * HashMap is empty.
+     *
+     * @return the ordered String of instructions
+     */
+    private String determineOrder() {
+        List<Character> availableStepsIds = new ArrayList<>();
+        List<Step> steps = getLocalCopyOfStepsMapAsArrayList();
+        StringBuilder order = new StringBuilder();
+
+        while (!steps.isEmpty()) {
+            steps.forEach((step) -> {
+                if (!step.hasPrevious() && !availableStepsIds.contains(step.id)) {
+                    availableStepsIds.add(step.id);
+                }
+            });
+
+            Collections.sort(availableStepsIds);
+            Character currentStepId = availableStepsIds.get(0);
+            order.append(currentStepId);
+
+            removeStepFromListAndPreviousSteps(steps, currentStepId);
+            availableStepsIds.remove(currentStepId);
+        }
+
+        return order.toString();
+    }
+
+    /**
+     * Primary method for Day 7, Part 2.
+     * <p>
+     * Determines the time it takes to complete all tasks with the number of workers given in {@link #workers}.
+     * For this purpose, {@link #stepsMap} is copied into a list. In a while loop, first the available steps are determined.
+     * Secondly, each available step is given to a worker, if one is available. The time it takes for the step to
+     * be completed is calculated in this step and stored with the respective ID. In the third step, the time left
+     * is reduced. If there is just one second left, the step is removed from the map of Steps, from the list of
+     * previous steps in each other step and from the list of steps in progress. Lastly the time taken is increased
+     * by 1. If there are no more steps left, the while loop ends and the total time taken is returned.
+     *
+     * @return time it takes to complete all the tasks
+     */
+    private int determineTime() {
```

Comment:
This of course is like a "doMagic()" method. Although I am sure that it does what it is supposed to do, it is way to long for somebody to understand who is reading the code for the first time. Not because it is not good to read and not well structured, only because of the underlying complexity of what it does. 

Also, consider what would happen if we refactored it? The test would break if we make a mistake, but we would not have a clue about why - or do we (also because we don't test it since it is a private method=?

To improve this, I would attempt to make it testable. For that, maybe moving it to a new class would be the first step, and splitting it into smaller, simpler methods. Then testing each of them in the new class.

If you feel that you have to test a private method, it has gotten too complex/large and something is wrong by design.

Reply:
You are right.. hmpfgrmbl. ;)